packages/gatsby-source-contentful/src/generate-schema.js

```diff
@@ -0,0 +1,318 @@
+    }
+  `)
+}
+
+export function generateSchema({
+  createTypes,
+  schema,
+  pluginConfig,
+  contentTypeItems,
+}) {
+  createTypes(`
+    interface ContentfulInternalReference implements Node {
+      contentful_id: String!
+      id: ID!
+    }
+  `)
+
+  createTypes(`
+    type ContentfulContentType implements Node {
+      id: ID!
+      name: String!
+      displayField: String!
+      description: String!
+    }
+  `)
+
+  createTypes(`
+    type ContentfulInternalSys @dontInfer {
+      type: String
+      revision: Int
+      contentType: ContentfulContentType @link(by: "id", from: "contentType___NODE")
+    }
+  `)
+
+  createTypes(`
+    interface ContentfulEntry implements Node @dontInfer {
+      contentful_id: String!
+      id: ID!
+      spaceId: String!
+      sys: ContentfulInternalSys
+    }
+  `)
+
+  generateAssetTypes({ createTypes })
+
+  // Contentful specific types
+  createTypes(
+    schema.buildObjectType({
+      name: `ContentfulNodeTypeRichText`,
+      fields: {
+        raw: {
+          type: `JSON`,
+          resolve(source) {
+            return source
+          },
+        },
+        references: {
+          type: `[ContentfulInternalReference]`,
+          resolve(source, args, context) {
+            const referencedEntries = new Set()
+            const referencedAssets = new Set()
+
+            // Locate all Contentful Links within the rich text data
+            // Traverse logic based on https://github.com/contentful/contentful-resolve-response
+            const traverse = obj => {
+              // eslint-disable-next-line guard-for-in
+              for (const k in obj) {
+                const v = obj[k]
+                if (v && v.sys && v.sys.type === `Link`) {
+                  if (v.sys.linkType === `Asset`) {
+                    referencedAssets.add(v.sys.id)
+                  }
+                  if (v.sys.linkType === `Entry`) {
+                    referencedEntries.add(v.sys.id)
+                  }
+                } else if (v && typeof v === `object`) {
+                  traverse(v)
+                }
+              }
+            }
+            traverse(source)
+
+            return context.nodeModel
+              .getAllNodes()
+              .filter(node =>
+                node.internal.owner === `gatsby-source-contentful` &&
+                node.internal.type === `ContentfulAsset`
+                  ? referencedAssets.has(node.contentful_id)
+                  : referencedEntries.has(node.contentful_id)
+              )
+          },
+        },
+      },
+      extensions: { dontInfer: {} },
+    })
+  )
+
+  createTypes(
+    schema.buildObjectType({
+      name: `ContentfulNodeTypeLocation`,
+      fields: {
+        lat: { type: `Float!` },
+        lon: { type: `Float!` },
+      },
+      extensions: {
+        dontInfer: {},
+      },
+    })
+  )
+
+  // @todo Is there a way to have this as string and let transformer-remark replace it with an object?
```

Comment:
Text fields that are transformed to markdown but actually don't use markdown are still the biggest node polluter. Any ideas around this?